[ONNX] Apply param_manipulation.py from onnx-script to op validation and dispatcher

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100554
* #105104
* __->__ #104679

Previous to this PR, we are comparing torch args/kwargs with OnnxFunction OpSchema without normalizing args/kwargs first. Essentially, the function signature is different between ATen and OnnxFunction, and onnx-script has preprocessing on these args/kwargs with an internal tools: `param_manipulation` for both eager mode and graph mode. This PR references on the internal tool to normalize the torch args/kwargs before feeding them to OnnxFunction during op_level_debug and dispatching. The PR significantly reduces the dispatching need on nearest matching mechanism.

titaiwangms committed:
[ONNX] Apply pparam_manipulation.py from onnx-script to op validation and dispatcher

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Apply param_manipulation.py from onnx-script to op validation and dispatcher"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Apply param_manipulation.py from onnx-script to op validation and dispatcher"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Apply param_manipulation.py from onnx-script to op validation and dispatcher"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Apply param_manipulation.py from onnx-script to op validation and dispatcher"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Apply param_manipulation.py from onnx-script to op validation and dispatcher"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Apply param_manipulation.py from onnx-script to op validation and dispatcher"


Previous to this PR, we are comparing torch args/kwargs with OnnxFunction OpSchema without normalizing args/kwargs first. Essentially, the function signature is different between ATen and OnnxFunction, and onnx-script has preprocessing on these args/kwargs with an internal tools: `param_manipulation` for both eager mode and graph mode. This PR references on the internal tool to normalize the torch args/kwargs before feeding them to OnnxFunction during op_level_debug and dispatching. The PR significantly reduces the dispatching need on nearest matching mechanism.

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -180,11 +179,8 @@ def filter_incompatible_and_dtype_convert_kwargs(kwargs):
-                    jit_type_utils.JitScalarType.from_dtype(value).onnx_type()
-                )
-            continue
+            elif isinstance(value, torch.dtype):
```
I think this part is overcoded. I will revert it back.

titaiwangms replied:
Reverted to else

titaiwangms commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -158,49 +157,21 @@ def forward(self, input, emb):
+            diagnostics.rules.find_opschema_matched_symbolic_function,
             diagnostics.levels.WARNING,
-            expected_error_node=expected_error_node,
+            expected_error_node="aten.new_zeros.default",
```
It works while it's input `size` is empty. Empty is not an acceptable dtpye in OnnxFunction. So it's not a perfect match. Besides this case, the other woule hit nearest match is the wrongly annotated cases.

titaiwangms commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -267,10 +234,103 @@ def _fx_args_to_torch_args(
+# NOTE: Referenced from onnxscript internal function.
+# Importing this function makes the code less robust, as it is not a public API on onnxscript.
+@_beartype.beartype
+def _tag_arguments_with_param_schemas(
```
Oh I see

titaiwangms replied:
Changed it to `_split_tangled_arguments_with_param_schemas`

titaiwangms committed:
Update on "[ONNX] Apply param_manipulation.py from onnx-script to op validation and dispatcher"


Previous to this PR, we are comparing torch args/kwargs with OnnxFunction OpSchema without normalizing args/kwargs first. Essentially, the function signature is different between ATen and OnnxFunction, and onnx-script has preprocessing on these args/kwargs with an internal tools: `param_manipulation` for both eager mode and graph mode. This PR references on the internal tool to normalize the torch args/kwargs before feeding them to OnnxFunction during op_level_debug and dispatching. The PR significantly reduces the dispatching need on nearest matching mechanism.

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -108,41 +89,32 @@ def validate_op_between_ort_torch(
+        except TypeError as type_error:
             diagnostic = diagnostic_context.inflight_diagnostic()
             diagnostic.with_additional_message(
                 f"### Op level debug is bypassed\n"
```
Addd

titaiwangms commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -170,7 +173,17 @@ def _find_the_perfect_or_nearest_match_onnxfunction(
+
+            # NOTE: OnnxFunction does not have the same function signature as the original
+            # PyTorch operator. We need to separate the input attributes from the arguments.
+            function_args, function_kwargs = _separate_input_attributes_from_arguments(
```
It would be a little weird, If I directly put this function inside `_OpSchemaWrapper`. So I changed the input to OnnxFunction. This way it is kind of extended to both schema usages. The function right now is called inside `find_perfect_match`.

titaiwangms commented on test/onnx/dynamo/test_registry_dispatcher.py:
```diff
@@ -316,10 +316,6 @@ def setUp(self):
-            common_utils.subtest(
-                ([torch.randn(3, 4), torch.randn(3, 4), torch.randn(3, 4)], {}, False),
-                name="non_perfect_match_due_to_too_many_inputs",
-            ),
```
https://github.com/pytorch/pytorch/issues/105095

titaiwangms committed:
Update on "[ONNX] Apply param_manipulation.py from onnx-script to op validation and dispatcher"


Previous to this PR, we are comparing torch args/kwargs with OnnxFunction OpSchema without normalizing args/kwargs first. Essentially, the function signature is different between ATen and OnnxFunction, and onnx-script has preprocessing on these args/kwargs with an internal tools: `param_manipulation` for both eager mode and graph mode. This PR references on the internal tool to normalize the torch args/kwargs before feeding them to OnnxFunction during op_level_debug and dispatching. The PR significantly reduces the dispatching need on nearest matching mechanism.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Apply param_manipulation.py from onnx-script to op validation and dispatcher"


Previous to this PR, we are comparing torch args/kwargs with OnnxFunction OpSchema without normalizing args/kwargs first. Essentially, the function signature is different between ATen and OnnxFunction, and onnx-script has preprocessing on these args/kwargs with an internal tools: `param_manipulation` for both eager mode and graph mode. This PR references on the internal tool to normalize the torch args/kwargs before feeding them to OnnxFunction during op_level_debug and dispatching. The PR significantly reduces the dispatching need on nearest matching mechanism.

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -460,42 +460,54 @@ def perfect_match_inputs(
                 return False
-        if set(kwargs) != self.attributes:
-            # If the attributes of the OpSchema and the kwargs don't match,
+        # TODO: attributes dtype is not supported right now, so we just check their existence.
```
Does https://github.com/pytorch/pytorch/issues/105095 address this TODO?

BowenBao commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -108,41 +89,32 @@ def validate_op_between_ort_torch(
-                f"\nBypass the test of running on ONNX Op {function_name} with "
-                f"ValueError: \n{value_error}.\n This is possibly raised by "
-                f"unsupported input args due to lack of Opschema.\n"
+            onnx_args, onnx_kwargs = _tag_arguments_with_param_schemas(
```
Q was for latter, thx.

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -419,21 +419,21 @@ def aten_new_full_dtype(self: TTensor, size: INT64, fill_value: TTensor, dtype:
 
-    def __init__(self, op_schema: onnx.defs.OpSchema):
+    def __init__(self, onnxfunction: "onnxscript.OnnxFunction"):
         """Initialize the OpSchemaWrapper.
```
This comment is also for the class name and class docstring above that github does not let me add comments. 
Better to update all names and docstrings, sry!

But I believe this is right and good change. Crux is in order to do input match, we also require onnxfunction param **ordering**. Such info is exclusively in `OnnxFunction.param_schemas()`, and not in `OpSchema`.

titaiwangms replied:
Use `_OnnxSchemaChecker`

BowenBao commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -267,10 +241,103 @@ def _fx_args_to_torch_args(
+# NOTE: Referenced from onnxscript internal function: _tag_arguments_with_param_schemas.
+# Importing this function makes the code less robust, as it is not a public API on onnxscript.
+@_beartype.beartype
+def _split_tangled_arguments_with_param_schemas(
```
synced offline, some clarification and renaming on terminologies is needed

titaiwangms replied:
Use `_convert_torch_args_to_onnxfunction_args`

titaiwangms committed:
Update on "[ONNX] Apply param_manipulation.py from onnx-script to op validation and dispatcher"


Previous to this PR, we are comparing torch args/kwargs with OnnxFunction OpSchema without normalizing args/kwargs first. Essentially, the function signature is different between ATen and OnnxFunction, and onnx-script has preprocessing on these args/kwargs with an internal tools: `param_manipulation` for both eager mode and graph mode. This PR references on the internal tool to normalize the torch args/kwargs before feeding them to OnnxFunction during op_level_debug and dispatching. The PR significantly reduces the dispatching need on nearest matching mechanism.

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
